sequencer/src/main/java/net/consensys/linea/rpc/services/LineaLimitedBundlePool.java

```diff
@@ -0,0 +1,192 @@
+      long blockNumber, PendingTransaction pendingTransaction) {
+    return getBundlesByBlockNumber(blockNumber).stream()
+        .filter(bundle -> bundle.pendingTransactions().contains(pendingTransaction))
+        .findAny();
+  }
+
+  /**
+   * Retrieves a TransactionBundle by its unique hash identifier.
+   *
+   * @param hash The hash identifier of the TransactionBundle.
+   * @return The TransactionBundle associated with the hash, or null if not found.
+   */
+  public TransactionBundle get(Hash hash) {
+    return cache.getIfPresent(hash);
+  }
+
+  /**
+   * Adds a new TransactionBundle to the cache and updates the block index.
+   *
+   * @param hash The hash identifier of the TransactionBundle.
+   * @param bundle The TransactionBundle to add.
+   */
+  public void put(Hash hash, TransactionBundle bundle) {
+    cache.put(hash, bundle);
+    addToBlockIndex(bundle);
+  }
+
+  /**
+   * Replaces an existing TransactionBundle in the cache and updates the block index.
+   *
+   * @param hash The hash identifier of the TransactionBundle.
+   * @param bundle The new TransactionBundle to replace the existing one.
+   */
+  public void replace(Hash hash, TransactionBundle bundle) {
+    TransactionBundle existing = cache.getIfPresent(hash);
+    if (existing != null) {
+      removeFromBlockIndex(existing);
+    }
+    cache.put(hash, bundle);
+    addToBlockIndex(bundle);
+  }
+
+  /**
+   * Removes all TransactionBundles associated with the given block number. First removes them from
+   * the block index, then removes them from the cache.
+   *
+   * @param blockNumber The block number whose bundles should be removed.
+   */
+  public void removeByBlockNumber(long blockNumber) {
+    List<TransactionBundle> bundles = blockIndex.remove(blockNumber);
+    if (bundles != null) {
+      for (TransactionBundle bundle : bundles) {
+        cache.invalidate(bundle.bundleIdentifier());
+      }
+    }
+  }
+
+  /**
+   * Adds a TransactionBundle to the block index.
+   *
+   * @param bundle The TransactionBundle to add.
+   */
+  private void addToBlockIndex(TransactionBundle bundle) {
+    long blockNumber = bundle.blockNumber().getValue();
+    blockIndex.computeIfAbsent(blockNumber, k -> new ArrayList<>()).add(bundle);
+  }
+
+  /**
+   * Removes a TransactionBundle from the block index.
+   *
+   * @param bundle The TransactionBundle to remove.
+   */
+  private void removeFromBlockIndex(TransactionBundle bundle) {
+    long blockNumber = bundle.blockNumber().getValue();
+    List<TransactionBundle> bundles = blockIndex.get(blockNumber);
+    if (bundles != null) {
+      bundles.remove(bundle);
+      if (bundles.isEmpty()) {
+        blockIndex.remove(blockNumber);
+      }
+    }
+  }
+
+  private int calculateWeight(TransactionBundle bundle) {
+    return bundle.pendingTransactions.stream()
+        .mapToInt(pt -> pt.getTransaction().getPayload().size())
+        .sum();
+  }
```

Comment:
`PendingTransaction` has a better way to compute the estimated memory size

Reply:
I used Transaction:getSize for now, since the PendingTransaction interface doesn't export memorySize()